### ISSUE TITLE
Identify error handling and database insertion mechanisms for WebCTRL, Egauge and HOBO

### ISSUE THREAD

**Issue by carlosparadis** [erdl/legacy-scrape-util#9]
Hi @jygh98 as we discussed, I think the next step concerning scrape-utils is looking for whatever error handling mechanisms Forrest implemented in his original script to see what we account for and what we don't on the data insertion. Even if we were not to have his code mapped to a Notebook, this is also necessary understanding (I believe) so we don't feel like we are stepping in glasses every time something goes wrong.

I know that Forrest set 2 kinds of errors mechanisms (maybe it was explained in his video), and this is actually how back then I realized that Python difference on the files when the server went down. 

You remember Apache Server, which handles the HTTP Requests? in `/var/log` some errors are registered. But this is something from Apache HTTP Server itself. I recall Forrest also had his way to handle errors, so it is this later error handling you should be more interested about.  

Lastly, the database insertion is part of his "inelegant solution", but we don't know exactly what that was, and we for sure may hit the same problem on the long run. The memory that I had of it was associated with the fact when all readings were submitted to the database, sometimes it would bouce back and it would create holes. Forrest then tried to send one reading at a time, but Postgres on our old version didn't accept that. We then were supposed to (I don't if we did or not) upgrade to a more recent version of Postgres, and he supposedly updated the script to update one reading at a time (Not sure). 

So on the database insertion side, the more you understand the better. If we can't understand the insertion, in the end, we just recreate it and see what happens, but not knowing what is going on, I believe, is worse.

**Comment by jygh98** [comment]
In the event of error during the psql insertion step, two functions are called to help document the error message and copy over the row into an error csv.  

The function mklog() creates a file "errlog.txt" in the /tmp/errors directory which keeps track of the error message that is produced during the script.  Do keep in mind that mklog is called whenever an error occurs throughout the entirety of the script, so it is not exclusively psql errors that is written but all including parsing errors.  An example of the error log is as followed:

    logtime............1499824757
    project............hobo-u12
    [Errno 13] Permission denied: '/var/local/scrape-util/hobo-u12/raws/10-10688123.csv'
    ...................endlog

The function errdata() essentially copies the row that creates error and rewrites them into an error-only csv called "psqlerr-timestamp.csv".  It doesn't appear that anything in scrape-util would read this           psqlerr-timestamp.csv and try re-insert it to the database, so what I am assuming was that Forrest would enter in these rows manually which was his "inelegant solution". 

All these errors files are written in the /tmp/errors folder in their respective project folder.  Also there may be files named "psqldups-timestamp.csv" and what these are a csv containing duplicated rows, we have the ability to change whether we want to save duplicates or not.  Lastly the two error functions, mklog() and errdata() can be found in the error_utils.py.

**Comment by carlosviansi** [comment]
@jygh98 Very nice detailed description of what is going on! Great job! :-) The line coding link is also very useful, wasn't aware of that. @ryantanaka take note of this one feature of Github, seeing we will be doing a lot of surgical modifications this may come in handy in our discussions! 

> A very interesting thing to note is the break statement at the end of the loop. I am not too confident on what is the purpose of it, as wouldn't all cases go through the else statement?

If I understand your question right, then no, it won't go in all cases of the else statement. Take note the [break](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L83) is not even inside the [else](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L80), but inside the [except](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L77). This means that the [break](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L83) is interrupting the [for loop](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L71). 

In other words, my understanding is that the break is to stop the entire insertion process if at least one row goes wrong. This is strange to me, as then I don't understand how we have holes in the data if the function supposedly halts at that point. Try looking on the behavior of the function from there onwards. 

```
con.set_session(autocommit=True)
```

This is it: the `autocommit`. He is ignoring the use of `transactions`. @ryantanaka Have a look on this above: What we keep hearing as the inelegant solution and the problems of one row failing all the other rows that were going to be inserted were addressed using this autocommit, as @jygh98 explained above. 

@jygh98 You should consider a TO-DO investigating if this auto-commit and disabling transactions were indeed the best option for our needs. @ryantanaka can tell you a bit more about what is a transaction in a database, or you can try googling and following-up with us here or in the office.  

For now, I would suggest you focus on better understanding the way Forrest handles the execution flow when a reading fails to update and the purpose of that break. 

On a small remark: You forgot to add the example `utf-8` error you showed me in the lab.

**Comment by carlosviansi** [comment]
https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L47

See the documentation of this function, it is the inelegant solution using the "throw it at the wall until it hits the hole" approach.

**Comment by jygh98** [comment]
So if the readings fails and is read as an error, the break would interrupt the for loop which stops the insertion process like you said.  But if autocommit is true, wouldn't that mean all the previous rows that were able to be read be uploaded since there are no transactions?

**Comment by carlosparadis** [comment]
You're correct. But I am not sure why you are asking this. The inelegant solution is still a solution: We were losing readings supposedly before it was put in place. 

From what I gather of this code, at the very least it will store the readings that are `duplicate on insertion` and that simple `failed to be inserted`.

The script inserts into a `reading` table in DHHL or frog_uhm, whose primary key is defined as follows:

reading_pkey" PRIMARY KEY, btree (purpose_id, reading_timestamp) 

So even the data you sent to Eileen if inserted by scrape-utils it will toss a duplicate error (or really anything since the constraint lives in the database). 

In scrape-util, the code stays stored in these dictionaries:

https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L58

https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/export/psql.py#L59

What I don't know is what happens to them afterward. Are they put somewhere to be re-attempted in the next time the cron job runs? Maybe you could try to trace that in the code using the figures of function calls I posted on slack. 

Also, I would recommend looking on the web for ways to handle streaming data into the database or whatever problem may arise from sending too many transactions to the database from sensor data.

**Comment by jygh98** [comment]
So I explained this earlier in the error propagation section but here it is below.

In the event of error during the psql insertion step, two functions are called to help document the error message and copy over the row into an error csv.  

The function [mklog()](https://github.com/erdl/legacy-scrape-util/blob/master/src/core/error_utils.py#L20) creates a file "errlog.txt" which keeps track of the error message that is produced during the script.  Do keep in mind that mklog is called whenever an error occurs throughout the entirety of the script, so it is not exclusively psql errors that is written but all including parsing errors.  An example of the error log is as followed:

    logtime............1499824757
    project............hobo-u12
    [Errno 13] Permission denied: '/var/local/scrape-util/hobo-u12/raws/10-10688123.csv'
    ...................endlog

The function [errdata()](https://github.com/erdl/legacy-scrape-util/blob/master/src/core/error_utils.py#L39) essentially copies the row that creates error and rewrites them into an error-only csv called "psqlerr-timestamp.csv".  It doesn't appear that anything in scrape-util would read this psqlerr-timestamp.csv and try re-insert it to the database, so what I am assuming was that Forrest would enter in these rows manually. 

All these errors files are written in the /var/local/scape-util/errors/ folder in their respective project folder.  Also there may be files named "psqldups-timestamp.csv" and what these are a csv containing duplicated rows, we have the ability to change whether we want to save duplicates or not.  Lastly the two error functions, mklog() and errdata() can be found in the error_utils.py.

TL:DR:  When error, scrape-util will make two files, one is a text file that keeps a log of the error and the other file is a csv that contains the rows that caused the error.  At this point it does not look like anything is called to try reinsert these rows, so I believe Forrest would try to manually enter in these rows.

**Comment by carlosparadis** [comment]
A reminder of to-dos to wrap this up we discussed today:

- Add the sample fro the errlog that is associated with UTC errors as a new post here. 
- Add the path and explain how we can make sense of the errorlog. There must be a format how errors are reported and about what. Let us know (I will move this to the wiki afterward). 

- Make a second post discussing how the current inelegant solution fares with how people store streaming data from sensors in relational databases (we need best practices).

**Comment by jygh98** [comment]
Sample error log from dhhl-hobo-u12
![image](https://user-images.githubusercontent.com/25787662/31638215-1e2f9ae0-b26e-11e7-8687-53ebe7475084.png)

The path for these error logs can be found in /var/local/scrape-util/errors.  In the log, the errors are formatted as seen above:
   logtime
   project
   error message

So I was talking to Ryan and Eileen about how we can possibly upload the data, Ryan suggested that we should instead just insert the raw data csv into a temporary loading table.  From there, we would have the script loop through the table to find distinct timestamps and put it into a new loading table.  If we have any duplicates, Eileen said she doesn't care about duplicates other than saving it like we are doing now.  So with this new loading table, each row will have an distinct timestamp and if we attempt to batch insert this table into the database it should work without issue.  With this it should make the insertion process much simpler and easier to trace in case of error.

**Comment by carlosparadis** [comment]
> So I was talking to Ryan and Eileen about how we can possibly upload the data, Ryan suggested that we should instead just insert the raw data csv into a temporary loading table. From there, we would have the script loop through the table to find distinct timestamps and put it into a new loading table. If we have any duplicates, Eileen said she doesn't care about duplicates other than saving it like we are doing now. So with this new loading table, each row will have a distinct timestamp and if we attempt to batch insert this table into the database it should work without issue. With this, it should make the insertion process much simpler and easier to trace in case of error.

@ryantanaka  @jygh98 I think you overlooked the question of storage which I wanted you to go after best practices on the internet. We are inserting `streaming` data as `hourly batch`, and there is no reason for this. It is indeed inconvenient at some points, where Eileen wanted to see the data immediately while visiting houses at the DHHL project. This also adds a bottleneck on the dashboards @kathrynparadis is creating: We will always see 1 hour later data. 

This is an interesting problem and one I am sure you would carry on what you learn elsewhere. Try to put a bit more googling and researching on it. It shouldn't absorb all your time in ERDL, but you can do quick searches here and there and add links here editing your posts as you go. Start by trying to look at why something like `TimeseriesDB` could or couldn't be a better way to handle this since it is an extension to `PostgresSQL`.

**Comment by jygh98** [comment]
I will be updating this response as I find possible solutions to our current implementation, but I would like to post these stackoverflow post where people are [encountering similar problems ](https://stackoverflow.com/questions/6025367/t-sql-deleting-all-duplicate-rows-but-keeping-one?noredirect=1&lq=1)to us with [duplicates in tables](https://stackoverflow.com/questions/33994259/sql-deleting-row-which-values-already-exist?noredirect=1&lq=1).

**Comment by carlosparadis** [comment]
We never followed up on the database insertion in the end, but [there was a test and permanent change to the cron job scrape-util requests being decreased to 5 minutes #19, which was handled well, so I believe we will not be touching this issue unless strictly required.

As far as the error documentation goes, it is currently being added as material to the wiki of the sensors project. See [this issue](https://github.com/erdl/sensors/issues/9) associated with said documentation.
